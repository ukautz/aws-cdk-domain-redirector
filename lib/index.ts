import * as cdk from 'aws-cdk-lib';
import {
  aws_certificatemanager as acm,
  aws_cloudfront as cloudfront,
  aws_cloudfront_origins as origins,
  aws_route53 as route53,
  aws_route53_targets as targets,
  aws_s3 as s3,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

const defaultCacheDuration = cdk.Duration.hours(1);

export interface DomainRedirectorProps {
  /**
   *
   * @default 1 hour
   */
  cacheDuration?: cdk.Duration;

  /**
   * A list of domains that are to be redirect to the target
   */
  domains: string[];

  /**
   * The Route53 hosted zone(s) of the domains.
   *
   * DNS validation records will be created within as well as the A records
   * that route the domains to the CloudFront
   */
  hostedZone: route53.IHostedZone;

  /**
   * The target domain to which the redirects will be delegated to
   */
  targetDomain: string;
}

/**
 * DomainRedirector implements an Level 3 CDK construct, whic
 *
 *  simple CloudFront and S3 based HTTP(S) redirection.
 *
 * Resource setup:
 * - S3 Bucket, that redirects all incoming requests to target domain
 * - CloudFront Distribution, that accepts requests for all (to-be-redirected) domains
 * - ACM Certificate, that validates all (to-be-redirected) domains and is associated with CloudFront Distribution
 *
 * @experimental
 */
export class DomainRedirector extends Construct {
  public readonly bucket: s3.IBucket;
  public readonly certificate: acm.ICertificate;
  public readonly distribution: cloudfront.IDistribution;

  constructor(scope: Construct, id: string, props: DomainRedirectorProps) {
    super(scope, id);

    if (props.domains.length === 0) throw new Error('No domains are provided');

    const apex = props.hostedZone.zoneName.replace(/\.+$/, '');
    const invalid = props.domains.filter((domain) => domain != apex && !domain.endsWith('.' + apex));
    if (invalid.length > 0) throw new Error(`Hosted zone name ${apex} does not match domains: ${invalid.join(', ')}`);

    this.bucket = this.initBucket(props);
    this.certificate = this.initCertificate(props);
    this.distribution = this.initDistribution(props);
    this.routeRecords(props);
  }

  private initBucket(props: DomainRedirectorProps): s3.Bucket {
    return new s3.Bucket(this, 'Bucket', {
      publicReadAccess: true,
      websiteRedirect: {
        hostName: props.targetDomain,
        protocol: s3.RedirectProtocol.HTTPS,
      },
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
  }

  private initCertificate(props: DomainRedirectorProps): acm.Certificate {
    return new acm.Certificate(this, 'Certificate', {
      domainName: props.domains[0],
      ...(props.domains.length > 1 ? { subjectAlternativeNames: props.domains.slice(1) } : {}),
      validation: acm.CertificateValidation.fromDns(props.hostedZone),
    });
  }

  private initDistribution(props: DomainRedirectorProps): cloudfront.Distribution {
    return new cloudfront.Distribution(this, 'Distribution', {
      domainNames: props.domains,
      certificate: this.certificate,
      defaultBehavior: {
        origin: new origins.S3Origin(this.bucket),
        cachePolicy: new cloudfront.CachePolicy(this, 'CachePolicy', {
          defaultTtl: props.cacheDuration ?? defaultCacheDuration,
        }),
      },
    });
  }

  private routeRecords(props: DomainRedirectorProps) {
    const target = route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(this.distribution));
    props.domains.forEach((domain) => {
      new route53.ARecord(this, `Record_${domain}`, {
        recordName: domain,
        zone: props.hostedZone,
        target,
      });
    });
  }
}
