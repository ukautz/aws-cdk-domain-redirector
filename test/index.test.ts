import * as cdk from 'aws-cdk-lib';
import { aws_certificatemanager as acm, aws_cloudfront as cloudfront, aws_route53 as route53 } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { type IConstruct } from 'constructs';
import { DomainRedirector, type DomainRedirectorProps } from '../lib';

/*
 * Example test
 */
describe('Domain Redirector', () => {
  describe('Intended setup', () => {
    // WHEN
    const stack = newTestStack();

    // THEN
    assertSnapshot(stack);
    assertRedirectBucket(stack);
    assertCloudFrontDistribution(stack);
    assertRecords(stack);
  });

  test('Domains not matching hosted zone fail', () => {
    expect(() => {
      newTestStack({
        domains: ['foo.tld', 'bar.tld'],
      });
    }).toThrowError();
  });
});

function getChild(parent: IConstruct, ...names: string[]): IConstruct {
  const child = parent.node.findChild(names.shift() as string);
  return names.length ? getChild(child, ...names) : child;
}

function assertRecords(stack: cdk.Stack): void {
  const template = Template.fromStack(stack);
  const distribution = getChild(stack, 'MyTestConstruct', 'Distribution') as cloudfront.Distribution;
  describe('Route53', () => {
    it('Creates Records', () => {
      template.resourceCountIs('AWS::Route53::RecordSet', 2);
    });
    it('Has record for www domain', () => {
      template.hasResourceProperties('AWS::Route53::RecordSet', {
        Name: 'www.domain.tld.',
        Type: 'A',
        AliasTarget: {
          DNSName: stack.resolve(distribution.domainName),
        },
      });
    });
    it('Has record for other domain', () => {
      template.hasResourceProperties('AWS::Route53::RecordSet', {
        Name: 'other.domain.tld.',
        Type: 'A',
        AliasTarget: {
          DNSName: stack.resolve(distribution.domainName),
        },
      });
    });
  });
}

function assertCloudFrontDistribution(stack: cdk.Stack): void {
  const template = Template.fromStack(stack);
  describe('CloudFront', () => {
    it('Creates distribution', () => {
      template.resourceCountIs('AWS::CloudFront::Distribution', 1);
    });
    it('Associates all domains', () => {
      template.hasResourceProperties('AWS::CloudFront::Distribution', {
        DistributionConfig: {
          Aliases: ['www.domain.tld', 'other.domain.tld'],
        },
      });
    });
    it('Has certificate associated', () => {
      const certificate = getChild(stack, 'MyTestConstruct', 'Certificate') as acm.Certificate;
      template.hasResourceProperties('AWS::CloudFront::Distribution', {
        DistributionConfig: {
          ViewerCertificate: {
            AcmCertificateArn: stack.resolve(certificate.certificateArn),
          },
        },
      });
    });
  });
}

function assertRedirectBucket(stack: cdk.Stack): void {
  const template = Template.fromStack(stack);
  describe('S3', () => {
    it('Creates bucket', () => {
      template.resourceCountIs('AWS::S3::Bucket', 1);
    });
    it('Forwards to target', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        WebsiteConfiguration: {
          RedirectAllRequestsTo: {
            HostName: 'target.tld',
            Protocol: 'https',
          },
        },
      });
    });
  });
}

function assertSnapshot(stack: cdk.Stack): void {
  const template = Template.fromStack(stack);
  it('Has consistent snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
}

function newTestStack(props?: Partial<DomainRedirectorProps>): cdk.Stack {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');
  const hostedZone = new route53.HostedZone(stack, 'HostedZone', {
    zoneName: 'domain.tld',
  });
  new DomainRedirector(stack, 'MyTestConstruct', {
    domains: ['www.domain.tld', 'other.domain.tld'],
    targetDomain: 'target.tld',
    ...props,
    hostedZone,
  });
  return stack;
}
