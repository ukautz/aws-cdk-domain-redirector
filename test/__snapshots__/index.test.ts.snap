// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Domain Redirector Intended setup Has consistent snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "HostedZoneDB99F866": {
      "Properties": {
        "Name": "domain.tld.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "MyTestConstructBucketC00BBB11": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "WebsiteConfiguration": {
          "RedirectAllRequestsTo": {
            "HostName": "target.tld",
            "Protocol": "https",
          },
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestConstructBucketPolicyE5FFA623": {
      "Properties": {
        "Bucket": {
          "Ref": "MyTestConstructBucketC00BBB11",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "MyTestConstructBucketC00BBB11",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "MyTestConstructCachePolicy03369574": {
      "Properties": {
        "CachePolicyConfig": {
          "DefaultTTL": 3600,
          "MaxTTL": 31536000,
          "MinTTL": 0,
          "Name": {
            "Fn::Join": [
              "",
              [
                "TestStackMyTestConstructCachePolicyD0D91B2C-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "ParametersInCacheKeyAndForwardedToOrigin": {
            "CookiesConfig": {
              "CookieBehavior": "none",
            },
            "EnableAcceptEncodingBrotli": false,
            "EnableAcceptEncodingGzip": false,
            "HeadersConfig": {
              "HeaderBehavior": "none",
            },
            "QueryStringsConfig": {
              "QueryStringBehavior": "none",
            },
          },
        },
      },
      "Type": "AWS::CloudFront::CachePolicy",
    },
    "MyTestConstructCertificate41948685": {
      "Properties": {
        "DomainName": "www.domain.tld",
        "DomainValidationOptions": [
          {
            "DomainName": "www.domain.tld",
            "HostedZoneId": {
              "Ref": "HostedZoneDB99F866",
            },
          },
          {
            "DomainName": "other.domain.tld",
            "HostedZoneId": {
              "Ref": "HostedZoneDB99F866",
            },
          },
        ],
        "SubjectAlternativeNames": [
          "other.domain.tld",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/MyTestConstruct/Certificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "MyTestConstructDistribution60637A64": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "www.domain.tld",
            "other.domain.tld",
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": {
              "Ref": "MyTestConstructCachePolicy03369574",
            },
            "Compress": true,
            "TargetOriginId": "TestStackMyTestConstructDistributionOrigin16656F6BD",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "http-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Fn::GetAtt": [
                          "MyTestConstructBucketC00BBB11",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "TestStackMyTestConstructDistributionOrigin16656F6BD",
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "MyTestConstructCertificate41948685",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructRecordotherdomaintldF7815C85": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "MyTestConstructDistribution60637A64",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "other.domain.tld.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MyTestConstructRecordwwwdomaintldA4BD1578": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "MyTestConstructDistribution60637A64",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "www.domain.tld.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
